File: ios/Classes/PerformanceGuardPlugin.swift
import Flutter
import UIKit
import Foundation

/// Main plugin class for iOS implementation
public class PerformanceGuardPlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    private var runLoopMonitor: RunLoopMonitor?
    private var isMonitoring = false
    private var options: PerformanceOptions?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = PerformanceGuardPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "com.example.performance_guard/methods",
            binaryMessenger: registrar.messenger()
        )
        instance.methodChannel = methodChannel
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "com.example.performance_guard/events",
            binaryMessenger: registrar.messenger()
        )
        instance.eventChannel = eventChannel

        let streamHandler = FreezeEventStreamHandler()
        eventChannel.setStreamHandler(streamHandler)
        instance.eventSink = streamHandler.eventSink
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENTS",
                                   message: "Invalid arguments",
                                   details: nil))
                return
            }
            options = PerformanceOptions(from: args)
            startMonitoring()
            result(nil)

        case "stop":
            stopMonitoring()
            result(nil)

        case "isMonitoring":
            result(isMonitoring)

        case "getConfiguration":
            result(options?.toDictionary())

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startMonitoring() {
        guard !isMonitoring, let opts = options else { return }

        isMonitoring = true

        runLoopMonitor = RunLoopMonitor(
            stallThresholdMs: opts.freezeThresholdMs,
            captureStackTraces: opts.captureStackTraces,
            onStallDetected: { [weak self] event in
                self?.sendFreezeEvent(event)
            }
        )
        runLoopMonitor?.start()
    }

    private func stopMonitoring() {
        runLoopMonitor?.stop()
        runLoopMonitor = nil
        isMonitoring = false
    }

    private func sendFreezeEvent(_ event: FreezeEvent) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(event.toDictionary())
        }
    }
}

/// Stream handler for freeze events
class FreezeEventStreamHandler: NSObject, FlutterStreamHandler {
    var eventSink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

/// Configuration options for iOS monitoring
struct PerformanceOptions {
    let freezeThresholdMs: Int
    let frameDropThresholdMs: Int
    let captureStackTraces: Bool
    let samplingIntervalMs: Int
    let includeAppState: Bool
    let minimumReportDurationMs: Int

    init(from dictionary: [String: Any]) {
        self.freezeThresholdMs = dictionary["freezeThresholdMs"] as? Int ?? 2000
        self.frameDropThresholdMs = dictionary["frameDropThresholdMs"] as? Int ?? 100
        self.captureStackTraces = dictionary["captureStackTraces"] as? Bool ?? true
        self.samplingIntervalMs = dictionary["samplingIntervalMs"] as? Int ?? 500
        self.includeAppState = dictionary["includeAppState"] as? Bool ?? true
        self.minimumReportDurationMs = dictionary["minimumReportDurationMs"] as? Int ?? 100
    }

    func toDictionary() -> [String: Any] {
        return [
            "freezeThresholdMs": freezeThresholdMs,
            "frameDropThresholdMs": frameDropThresholdMs,
            "captureStackTraces": captureStackTraces,
            "samplingIntervalMs": samplingIntervalMs,
            "includeAppState": includeAppState,
            "minimumReportDurationMs": minimumReportDurationMs
        ]
    }
}

/// Freeze event data structure
struct FreezeEvent {
    let freezeDurationMs: Int
    let timestamp: Int64
    let stackTrace: String?
    let threadName: String
    let platform: String
    let type: String
    let metadata: [String: Any]?

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "freezeDurationMs": freezeDurationMs,
            "timestamp": timestamp,
            "threadName": threadName,
            "platform": platform,
            "type": type
        ]
        if let stackTrace = stackTrace {
            dict["stackTrace"] = stackTrace
        }
        if let metadata = metadata {
            dict["metadata"] = metadata
        }
        return dict
    }
}

File: android/src/main/kotlin/com/example/performance_guard/PerformanceGuardPlugin.kt
package com.example.performance_guard

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*

/** Main plugin class for Android implementation */
class PerformanceGuardPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var anrWatchdog: AnrWatchdog? = null
    private var choreographerMonitor: ChoreographerMonitor? = null

    private val mainScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private val ioScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private var isMonitoring = false
    private var options: PerformanceOptions? = null

    companion object {
        const val METHOD_CHANNEL = "com.example.performance_guard/methods"
        const val EVENT_CHANNEL = "com.example.performance_guard/events"
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        stopMonitoring()
        mainScope.cancel()
        ioScope.cancel()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> {
                val optionsMap = call.arguments<Map<String, Any>>() ?: emptyMap()
                options = PerformanceOptions.fromMap(optionsMap)
                startMonitoring()
                result.success(null)
            }
            "stop" -> {
                stopMonitoring()
                result.success(null)
            }
            "isMonitoring" -> {
                result.success(isMonitoring)
            }
            "getConfiguration" -> {
                result.success(options?.toMap())
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun startMonitoring() {
        if (isMonitoring) return

        val opts = options ?: return

        isMonitoring = true

        // Start ANR watchdog on IO dispatcher
        ioScope.launch {
            anrWatchdog = AnrWatchdog(
                freezeThresholdMs = opts.freezeThresholdMs,
                captureStackTraces = opts.captureStackTraces,
                onFreezeDetected = ::sendFreezeEvent
            )
            anrWatchdog?.start()
        }

        // Start Choreographer monitor on main dispatcher
        mainScope.launch {
            choreographerMonitor = ChoreographerMonitor(
                frameDropThresholdMs = opts.frameDropThresholdMs,
                onFrameDropDetected = ::sendFreezeEvent
            )
            choreographerMonitor?.start()
        }
    }

    private fun stopMonitoring() {
        anrWatchdog?.stop()
        choreographerMonitor?.stop()
        anrWatchdog = null
        choreographerMonitor = null
        isMonitoring = false
    }

    private fun sendFreezeEvent(event: FreezeEvent) {
        mainScope.launch {
            eventSink?.success(event.toMap())
        }
    }
}

/** Configuration options for Android monitoring */
data class PerformanceOptions(
    val freezeThresholdMs: Int,
    val frameDropThresholdMs: Int,
    val captureStackTraces: Boolean,
    val samplingIntervalMs: Int,
    val includeAppState: Boolean,
    val minimumReportDurationMs: Int
) {
    companion object {
        fun fromMap(map: Map<String, Any>): PerformanceOptions {
            return PerformanceOptions(
                freezeThresholdMs = map["freezeThresholdMs"] as? Int ?: 2000,
                frameDropThresholdMs = map["frameDropThresholdMs"] as? Int ?: 100,
                captureStackTraces = map["captureStackTraces"] as? Boolean ?: true,
                samplingIntervalMs = map["samplingIntervalMs"] as? Int ?: 500,
                includeAppState = map["includeAppState"] as? Boolean ?: true,
                minimumReportDurationMs = map["minimumReportDurationMs"] as? Int ?: 100
            )
        }
    }

    fun toMap(): Map<String, Any> {
        return mapOf(
            "freezeThresholdMs" to freezeThresholdMs,
            "frameDropThresholdMs" to frameDropThresholdMs,
            "captureStackTraces" to captureStackTraces,
            "samplingIntervalMs" to samplingIntervalMs,
            "includeAppState" to includeAppState,
            "minimumReportDurationMs" to minimumReportDurationMs
        )
    }
}

/** Freeze event data class */
data class FreezeEvent(
    val freezeDurationMs: Int,
    val timestamp: Long,
    val stackTrace: String?,
    val threadName: String,
    val platform: String = "android",
    val type: String,
    val metadata: Map<String, Any>? = null
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>(
            "freezeDurationMs" to freezeDurationMs,
            "timestamp" to timestamp,
            "threadName" to threadName,
            "platform" to platform,
            "type" to type
        )
        stackTrace?.let { map["stackTrace"] = it }
        metadata?.let { map["metadata"] = it }
        return map
    }
}
